server/webapp/WEB-INF/rails.new/spec/controllers/api_v1/admin/config_repos_controller_spec.rb

```diff
@@ -0,0 +1,433 @@
+      end
+
+      it 'should render 404 when a package does not exist' do
+        @config_repo_service.stub(:getConfigRepo).with('non-existent-package-id').and_return(nil)
+        put_with_api_header :update, id: 'non-existent-package-id'
+        expect(response).to have_api_message_response(404, 'Either the resource you requested was not found, or you are not authorized to perform this action.')
+      end
+      
+    end
+
+    describe :security do
+      before(:each) do
+        @config_repo_service.stub(:getConfigRepo).and_return(@config_repo_id)
+        controller.stub(:check_for_stale_request).and_return(nil)
+      end
+
+      it 'should allow anyone, with security disabled' do
+        disable_security
+        expect(controller).to allow_action(:put, :update, id: @config_repo_id)
+      end
+
+      it 'should disallow anonymous users, with security enabled' do
+        enable_security
+        login_as_anonymous
+        expect(controller).to disallow_action(:put, :update, id: @config_repo_id).with(401, 'You are not authorized to perform this action.')
+      end
+
+      it 'should disallow normal users, with security enabled' do
+        login_as_user
+        expect(controller).to disallow_action(:put, :update, id: @config_repo_id).with(401, 'You are not authorized to perform this action.')
+      end
+
+      it 'should allow admin users, with security enabled' do
+        login_as_admin
+        expect(controller).to allow_action(:put, :update, id: @config_repo_id)
+      end
+
+    end
+
+    describe :route do
+      describe :with_header do
+        it 'should route to update action of config repo controller for specified package id' do
+          expect(:put => 'api/admin/config_repos/foo123').to route_to(action: 'update', controller: 'api_v1/admin/config_repos', id: 'foo123')
+        end
+        it 'should route to update action of config repo controller for package_id with dots' do
+          expect(:put => 'api/admin/config_repos/foo.bar').to route_to(action: 'update', controller: 'api_v1/admin/config_repos', id: 'foo.bar')
+        end
+      end
+      describe :without_header do
+        before :each do
+          teardown_header
+        end
+        it 'should not route to update action of packages controller without header' do
+          expect(:put => 'api/admin/config_repos/foo').to_not route_to(put: 'update', controller: 'api_v1/admin/config_repos')
+          expect(:put => 'api/admin/config_repos/foo').to route_to(controller: 'application', action: 'unresolved', url: 'api/admin/config_repos/foo')
+        end
+      end
+    end
+  end
+
+  def get_config_repo_json(repo_id)
+    {
+      :id => repo_id,
+      :plugin => 'config-repo-json-plugin',
+      :material => {
+        :type => 'git',
+        :attributes => {
+          :url => 'git://foo',
+          :destination => nil,
+          :filter => nil,
+          :invert_filter => false,
+          :name => nil,
+          :auto_update => true,
```

Comment:
[this](https://github.com/gocd/gocd/blob/master/config/config-api/src/com/thoughtworks/go/config/remote/ConfigRepoConfig.java#L174-L177) one